File: apps/server/core/src/test/java/io/github/rohingosling/eca/core/EvaluatorTest.java
//---------------------------------------------------------------------------------------------------------------------
// Project: ECA Rule Engine
// Version: 1.0
// Date:    2022
// Author:  Rohin Gosling
//
// Description:
//
//   Example and trace tests for the pure stateless evaluator.
//
// TODO:
//
//   1. None.
//
//---------------------------------------------------------------------------------------------------------------------

package io.github.rohingosling.eca.core;

import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.assertj.core.api.Assertions.catchThrowableOfType;

//*********************************************************************************************************************
// Class: EvaluatorTest
//
// Description:
//
//   Example and trace tests for the pure stateless evaluator.
//
//*********************************************************************************************************************

class EvaluatorTest
{
    //=================================================================================================================
    // Fields
    //=================================================================================================================

    private final Evaluator evaluator = new Evaluator ();
    private final CompiledModel model = CoreFixtures.referenceModel ();

    //=================================================================================================================
    // Methods
    //=================================================================================================================

    //-----------------------------------------------------------------------------------------------------------------
    // Method: absentAndEmptyPayloadsHaveIdenticalSemantics
    //
    // Description:
    //
    //   Performs the absent and empty payloads have identical semantics operation using the supplied inputs and
    //   current state.
    //
    //-----------------------------------------------------------------------------------------------------------------

    @Test
    void absentAndEmptyPayloadsHaveIdenticalSemantics ()
    {
        EvaluationResult absentResult = this.evaluator.evaluate
        (
            this.model, EventOccurrence.withoutPayload ( "signal.received" ), true
        );
        EvaluationResult emptyResult = this.evaluator.evaluate
        (
            this.model, new EventOccurrence ( "signal.received", Collections.emptyMap () ), true
        );

        assertThat ( absentResult ).isEqualTo ( emptyResult );
        assertThat ( absentResult.getSelectedActionId () ).contains ( "record" );
    }

    //-----------------------------------------------------------------------------------------------------------------
    // Method: multipleMatchingRulesForTheSameActionReturnThatAction
    //
    // Description:
    //
    //   Performs the multiple matching rules for the same action return that action operation using the supplied
    //   inputs and current state.
    //
    //-----------------------------------------------------------------------------------------------------------------

    @Test
    void multipleMatchingRulesForTheSameActionReturnThatAction ()
    {
        EvaluationResult result = this.evaluator.evaluate ( this.model, CoreFixtures.signal ( 12L ), true );

        assertThat ( result.getSelectedActionId () ).contains ( "record" );
        assertThat ( result.getTrace ().orElseThrow ().getRules () )
            .extracting ( RuleTrace::getMatched )
            .containsOnly ( true );
    }

    //-----------------------------------------------------------------------------------------------------------------
    // Method: multipleMatchingDistinctActionsProduceAnExplicitAmbiguityFailure
    //
    // Description:
    //
    //   Performs the multiple matching distinct actions produce an explicit ambiguity failure operation using the
    //   supplied inputs and current state.
    //
    //-----------------------------------------------------------------------------------------------------------------

    @Test
    void multipleMatchingDistinctActionsProduceAnExplicitAmbiguityFailure ()
    {
        assertThatThrownBy
        (
            () -> this.evaluator.evaluate
            (
                CoreFixtures.referenceModelWithAmbiguousRule (), CoreFixtures.signal ( 12L ), true
            )
        ).isInstanceOf ( EvaluationInputException.class )
            .hasMessage
            (
                "Event 'signal.received' matched distinct actions [alert, record] through rules "
                    + "[property.alert, record.high.1, record.high.2, record.signal]."
            )
            .hasFieldOrPropertyWithValue ( "code", "ambiguous-action-selection" )
            .hasFieldOrPropertyWithValue ( "pointer", "/type" );
    }

    //-----------------------------------------------------------------------------------------------------------------
    // Method: missingDependenciesAreFalseAndAppearInCanonicalTrace
    //
    // Description:
    //
    //   Verifies that missing dependencies are false and appear in canonical trace and fails the test when the
    //   observed behavior differs.
    //
    //-----------------------------------------------------------------------------------------------------------------

    @Test
    void missingDependenciesAreFalseAndAppearInCanonicalTrace ()
    {
        EvaluationResult result = this.evaluator.evaluate ( this.model, CoreFixtures.signal ( null ), true );
        EvaluationTrace trace = result.getTrace ().orElseThrow ();

        assertThat ( trace.getConditions () ).extracting ( ConditionTrace::getConditionId )
            .containsExactly ( "always", "level.high" );
        assertThat ( trace.getConditions ().get ( 1 ).getResult () ).isFalse ();
        assertThat ( trace.getConditions ().get ( 1 ).getReason () ).isEqualTo ( TraceReason.MISSING_DEPENDENCY );
        assertThat ( trace.getConditions ().get ( 1 ).getMissingDependencies () ).containsExactly ( "level" );
        assertThat ( trace.getRules () ).extracting ( RuleTrace::getRuleId )
            .containsExactly ( "record.high.1", "record.high.2", "record.signal" );
    }

    //-----------------------------------------------------------------------------------------------------------------
    // Method: unknownEventSelectsNothingWithoutFailure
    //
    // Description:
    //
    //   Performs the unknown event selects nothing without failure operation using the supplied inputs and current
    //   state.
    //
    //-----------------------------------------------------------------------------------------------------------------

    @Test
    void unknownEventSelectsNothingWithoutFailure ()
    {
        EvaluationResult result = this.evaluator.evaluate
        (
            this.model, EventOccurrence.withoutPayload ( "signal.unknown" ), true
        );

        assertThat ( result.getSelectedAction () ).isEmpty ();
        assertThat ( result.getTrace ().orElseThrow ().getConditions () ).isEmpty ();
        assertThat ( result.getTrace ().orElseThrow ().getRules () ).isEmpty ();
    }

    //-----------------------------------------------------------------------------------------------------------------
    // Method: knownEventRejectsUndeclaredOrIncompatiblePayloadValues
    //
    // Description:
    //
    //   Performs the known event rejects undeclared or incompatible payload values operation using the supplied inputs
    //   and current state.
    //
    //-----------------------------------------------------------------------------------------------------------------

    @Test
    void knownEventRejectsUndeclaredOrIncompatiblePayloadValues ()
    {
        assertThatThrownBy
        (
            () -> this.evaluator.evaluate
            (
                this.model,
                new EventOccurrence
                (
                    "signal.received", Collections.singletonMap ( "unexpected", Value.booleanValue ( true ) )
                ),
                false
            )
        ).isInstanceOf ( EvaluationInputException.class )
            .hasFieldOrPropertyWithValue ( "code", "undeclared-event-parameter" )
            .hasFieldOrPropertyWithValue ( "pointer", "/payload/unexpected" );

        assertThatThrownBy
        (
            () -> this.evaluator.evaluate
            (
                this.model,
                new EventOccurrence
                (
                    "signal.received", Collections.singletonMap ( "level", Value.stringValue ( "high" ) )
                ),
                false
            )
        ).isInstanceOf ( EvaluationInputException.class )
            .hasFieldOrPropertyWithValue ( "code", "incompatible-event-parameter" )
            .hasFieldOrPropertyWithValue ( "pointer", "/payload/level" );
    }

    //-----------------------------------------------------------------------------------------------------------------
    // Method: traceGenerationDoesNotChangeTheSelectedAction
    //
    // Description:
    //
    //   Performs the trace generation does not change the selected action operation using the supplied inputs and
    //   current state.
    //
    //-----------------------------------------------------------------------------------------------------------------

    @Test
    void traceGenerationDoesNotChangeTheSelectedAction ()
    {
        EvaluationResult tracedResult = this.evaluator.evaluate ( this.model, CoreFixtures.signal ( 12L ), true );
        EvaluationResult untracedResult = this.evaluator.evaluate ( this.model, CoreFixtures.signal ( 12L ), false );

        assertThat ( tracedResult.getSelectedActionId () ).isEqualTo ( untracedResult.getSelectedActionId () );
        assertThat ( untracedResult.getTrace () ).isEmpty ();
    }

    //-----------------------------------------------------------------------------------------------------------------
    // Method: ruleOrderPermutationPreservesTheSuccessfulEvaluation
    //
    // Description:
    //
    //   Performs the rule order permutation preserves the successful evaluation operation using the supplied inputs
    //   and current state.
    //
    //-----------------------------------------------------------------------------------------------------------------

    @Test
    void ruleOrderPermutationPreservesTheSuccessfulEvaluation ()
    {
        EventOccurrence eventOccurrence = CoreFixtures.signal ( 12L );
        EvaluationResult originalResult = this.evaluator.evaluate
        (
            CoreFixtures.referenceModel (), eventOccurrence, true
        );
        EvaluationResult permutedResult = this.evaluator.evaluate
        (
            CoreFixtures.referenceModelWithReversedRules (), eventOccurrence, true
        );

        assertThat ( permutedResult ).isEqualTo ( originalResult );
    }

    //-----------------------------------------------------------------------------------------------------------------
    // Method: ruleOrderPermutationPreservesTheAmbiguityDiagnostic
    //
    // Description:
    //
    //   Performs the rule order permutation preserves the ambiguity diagnostic operation using the supplied inputs and
    //   current state.
    //
    //-----------------------------------------------------------------------------------------------------------------

    @Test
    void ruleOrderPermutationPreservesTheAmbiguityDiagnostic ()
    {
        EventOccurrence eventOccurrence = CoreFixtures.signal ( 12L );
        EvaluationInputException originalException = catchThrowableOfType
        (
            EvaluationInputException.class,
            () -> this.evaluator.evaluate
            (
                CoreFixtures.referenceModelWithAmbiguousRule (), eventOccurrence, false
            )
        );
        EvaluationInputException permutedException = catchThrowableOfType
        (
            EvaluationInputException.class,
            () -> this.evaluator.evaluate
            (
                CoreFixtures.referenceModelWithAmbiguousRuleAndReversedRules (), eventOccurrence, false
            )
        );

        assertThat ( permutedException.getCode () ).isEqualTo ( originalException.getCode () );
        assertThat ( permutedException.getMessage () ).isEqualTo ( originalException.getMessage () );
        assertThat ( permutedException.getPointer () ).isEqualTo ( originalException.getPointer () );
    }

    //-----------------------------------------------------------------------------------------------------------------
    // Method: largeAmbiguityMessagesAreBoundedDeterministicAndOrderInvariant
    //
    // Description:
    //
    //   Verifies that large ambiguity messages are bounded deterministic and order invariant and fails the test when
    //   the observed behavior differs.
    //
    //-----------------------------------------------------------------------------------------------------------------

    @Test
    void largeAmbiguityMessagesAreBoundedDeterministicAndOrderInvariant ()
    {
        EventOccurrence eventOccurrence = EventOccurrence.withoutPayload ( "signal.received" );
        CompiledModel originalModel = this.createLargeAmbiguousModel ( false );
        CompiledModel permutedModel = this.createLargeAmbiguousModel ( true );
        EvaluationInputException originalException = catchThrowableOfType
        (
            EvaluationInputException.class,
            () -> this.evaluator.evaluate ( originalModel, eventOccurrence, false )
        );
        EvaluationInputException repeatedException = catchThrowableOfType
        (
            EvaluationInputException.class,
            () -> this.evaluator.evaluate ( originalModel, eventOccurrence, false )
        );
        EvaluationInputException permutedException = catchThrowableOfType
        (
            EvaluationInputException.class,
            () -> this.evaluator.evaluate ( permutedModel, eventOccurrence, false )
        );
        String message = originalException.getMessage ();

        assertThat ( message.codePointCount ( 0, message.length () ) )
            .isEqualTo ( DiagnosticMessagePolicy.MAXIMUM_MESSAGE_CODE_POINTS );
        assertThat ( message ).endsWith ( "\u2026" );
        assertThat ( repeatedException.getMessage () ).isEqualTo ( message );
        assertThat ( permutedException.getMessage () ).isEqualTo ( message );
    }

    //-----------------------------------------------------------------------------------------------------------------
    // Method: createLargeAmbiguousModel
    //
    // Description:
    //
    //   Constructs large ambiguous model from the supplied inputs without mutating the caller's source values.
    //
    // Arguments:
    //
    //   reverseDeclarationOrder (boolean):
    //     The reverse declaration order used by this operation.
    //
    // Returns:
    //
    //   The newly constructed value, model element, or immutable state projection.
    //
    //-----------------------------------------------------------------------------------------------------------------

    private CompiledModel createLargeAmbiguousModel ( boolean reverseDeclarationOrder )
    {
        EventType eventType = new EventType
        (
            "signal.received", "Signal received", Collections.emptyList ()
        );
        Condition condition = new Condition
        (
            "always", "Always", Collections.emptyList (), PredicateDefinition.always ()
        );
        List<Action> actions = new ArrayList<> ();
        List<Rule> rules     = new ArrayList<> ();

        for ( int index = 0; index < 17; index++ )
        {
            String indexText        = Integer.toString ( index );
            String indexSegment     = "0".repeat ( 3 - indexText.length () ) + indexText;
            String actionPrefix     = "action." + indexSegment + ".";
            String rulePrefix       = "rule." + indexSegment + ".";
            String actionIdentifier = actionPrefix + "a".repeat ( 128 - actionPrefix.length () );
            String ruleIdentifier   = rulePrefix + "r".repeat ( 128 - rulePrefix.length () );

            actions.add ( new Action ( actionIdentifier, "Action " + indexText, Collections.emptyMap () ) );
            rules.add
            (
                new Rule
                (
                    ruleIdentifier,
                    "Rule " + indexText,
                    eventType.getId (),
                    condition.getId (),
                    actionIdentifier
                )
            );
        }

        if ( reverseDeclarationOrder )
        {
            Collections.reverse ( actions );
            Collections.reverse ( rules );
        }

        // Return the newly constructed CompiledModel instance.

        return new CompiledModel
        (
            "1.0",
            "large-ambiguity",
            Collections.singletonList ( eventType ),
            Collections.singletonList ( condition ),
            actions,
            rules
        );
    }
}

File: apps/client/src/engine/evaluator.test.ts
//---------------------------------------------------------------------------------------------------------------------
// Project: ECA Rule Engine Laboratory
// Version: 0.1.0
// Date:    2022
// Author:  Rohin Gosling
//
// Description:
//
//   Focused version 1 predicate edge cases beyond the shared conformance corpus.
//
//---------------------------------------------------------------------------------------------------------------------

import { describe, expect, it } from "vitest";
import { MAXIMUM_DIAGNOSTIC_MESSAGE_CODE_POINTS } from "../contracts/contract-limits";
import type { StatelessECAModel } from "../contracts/model.generated";
import { comparePredicate, evaluate, jsonValuesEqual, limitDiagnosticMessage } from "./evaluator";

//---------------------------------------------------------------------------------------------------------------------
// Function: createLongIdentifier
//
// Description:
//
//   Constructs long identifier from the supplied inputs without mutating the caller's source values.
//
// Arguments:
//
//   prefix (string):
//     The prefix used by this operation.
//
//   index (number):
//     The index used by this operation.
//
//   filler (string):
//     The filler used by this operation.
//
// Returns:
//
//   The newly constructed value, model element, or immutable state projection.
//
//---------------------------------------------------------------------------------------------------------------------

function createLongIdentifier ( prefix: string, index: number, filler: string ): string
{
    const identifierPrefix = `${ prefix }.${ index.toString ().padStart ( 3, "0" ) }.`;

    // Return the value produced by this code path.

    return identifierPrefix + filler.repeat ( 128 - identifierPrefix.length );
}

//---------------------------------------------------------------------------------------------------------------------
// Function: createLargeAmbiguousModel
//
// Description:
//
//   Constructs large ambiguous model from the supplied inputs without mutating the caller's source values.
//
// Arguments:
//
//   reverseDeclarationOrder (boolean):
//     The reverse declaration order used by this operation.
//
// Returns:
//
//   The newly constructed value, model element, or immutable state projection.
//
//---------------------------------------------------------------------------------------------------------------------

function createLargeAmbiguousModel ( reverseDeclarationOrder: boolean ): StatelessECAModel
{
    const indexes = Array.from
    (
        {
            length: 17
        }, ( _, index ) => index
    );
    const actions = indexes.map
    (
        index => (
        {
            id: createLongIdentifier ( "action", index, "a" ),
            name: `Action ${ index }`,
            parameters:
                {
                }
        } )
    );
    const rules = indexes.map
    (
        index => (
        {
            id: createLongIdentifier ( "rule", index, "r" ),
            name: `Rule ${ index }`,
            event: "signal.received",
            condition: "always",
            action: actions [ index ].id
        } )
    );

    // Return the value selected by the evaluated condition.

    return (
        {
            schemaVersion: "1.0",
            id: "large-ambiguity",
            name: "Large ambiguity",
            parameters: [],
            payloads: [],
            events: [
                {
                    id: "signal.received", name: "Signal received"
                } ],
            conditions: [
                {
                    id: "always", name: "Always", dependencies: [], predicate:
                    {
                        name: "always", arguments:
                        {
                        }
                    }
                } ],
            actions: reverseDeclarationOrder ? actions.reverse () : actions,
            rules: reverseDeclarationOrder ? rules.reverse () : rules
        }
    );
}

//---------------------------------------------------------------------------------------------------------------------
// Test Suite: version 1 predicate semantics
//
// Description:
//
//   Verifies version 1 predicate semantics and records the expected externally observable behavior for future changes.
//
//---------------------------------------------------------------------------------------------------------------------
describe
(
    "version 1 predicate semantics", () =>
    {
        //-----------------------------------------------------------------------------------------------------------------
        // Test: compares JSON objects structurally and treats negative zero as zero
        //
        // Description:
        //
        //   Verifies compares JSON objects structurally and treats negative zero as zero and records the expected
        //   externally observable behavior for future changes.
        //
        //-----------------------------------------------------------------------------------------------------------------
        it
        (
            "compares JSON objects structurally and treats negative zero as zero", () =>
            {
                expect
                (
                    jsonValuesEqual
                    (
                        {
                            first: 1, second: [ true ]
                        },
                        {
                            second: [ true ], first: 1
                        }
                    )
                ).toBe ( true );
                expect ( jsonValuesEqual ( -0, 0 ) ).toBe ( true );
            }
        );

        //-----------------------------------------------------------------------------------------------------------------
        // Test: uses Unicode code-point ordering for strings
        //
        // Description:
        //
        //   Verifies uses Unicode code-point ordering for strings and records the expected externally observable behavior
        //   for future changes.
        //
        //-----------------------------------------------------------------------------------------------------------------
        it
        (
            "uses Unicode code-point ordering for strings", () =>
            {
                expect ( comparePredicate ( "greaterThan", "😀", "\uE000" ) ).toBe ( true );
                expect ( comparePredicate ( "lessThan", "\uE000", "😀" ) ).toBe ( true );
            }
        );

        //-----------------------------------------------------------------------------------------------------------------
        // Test: supports structural array membership and string substrings
        //
        // Description:
        //
        //   Verifies supports structural array membership and string substrings and records the expected externally
        //   observable behavior for future changes.
        //
        //-----------------------------------------------------------------------------------------------------------------
        it
        (
            "supports structural array membership and string substrings", () =>
            {
                expect
                (
                    comparePredicate
                    (
                        "contains", [
                        {
                            value: 7
                        } ],
                        {
                            value: 7
                        }
                    )
                ).toBe ( true );
                expect ( comparePredicate ( "contains", "stateless engine", "engine" ) ).toBe ( true );
            }
        );

        //-----------------------------------------------------------------------------------------------------------------
        // Test: treats an omitted prototype-named dependency as missing
        //
        // Description:
        //
        //   Verifies treats an omitted prototype-named dependency as missing and records the expected externally
        //   observable behavior for future changes.
        //
        //-----------------------------------------------------------------------------------------------------------------
        it
        (
            "treats an omitted prototype-named dependency as missing", () =>
            {
                const model: StatelessECAModel =
                {
                    schemaVersion: "1.0",
                    id: "prototype-dependency-model",
                    name: "Prototype dependency model",
                    parameters: [
                    {
                        id: "toString", name: "Text", type: "string"
                    } ],
                    payloads: [
                    {
                        id: "event-payload", name: "Event payload", parameters: [ "toString" ]
                    } ],
                    events: [
                    {
                        id: "event", name: "Event", payload: "event-payload"
                    } ],
                    conditions: [
                    {
                        id: "condition", name: "Condition", dependencies: [ "toString" ],
                        predicate:
                        {
                            name: "notEquals", arguments:
                            {
                                parameter: "toString", value: "present"
                            }
                        }
                    } ],
                    actions: [
                    {
                        id: "action", name: "Action", parameters:
                        {
                        }
                    } ],
                    rules: [
                    {
                        id: "rule", name: "Rule", event: "event", condition: "condition", action: "action"
                    } ]
                };

                const result = evaluate
                (
                    {
                        model, event:
                        {
                            type: "event", payload:
                            {
                            }
                        }, includeTrace: true
                    }
                );

                expect ( result ).toMatchObject
                (
                    {
                        outcome: "no-action", selectedAction: null
                    }
                );
                expect ( result.trace?.conditions ).toEqual
                (
                    [
                    {
                        condition: "condition", result: false, reason: "missing-dependency", missingDependencies: [ "toString" ]
                    } ]
                );
            }
        );

        //-----------------------------------------------------------------------------------------------------------------
        // Test: returns one action when several matching rules agree on that action
        //
        // Description:
        //
        //   Verifies returns one action when several matching rules agree on that action and records the expected
        //   externally observable behavior for future changes.
        //
        //-----------------------------------------------------------------------------------------------------------------
        it
        (
            "returns one action when several matching rules agree on that action", () =>
            {
                const model: StatelessECAModel =
                {
                    schemaVersion: "1.0",
                    id: "agreeing-rules-model",
                    name: "Agreeing rules model",
                    parameters: [],
                    payloads: [],
                    events: [
                    {
                        id: "event", name: "Event"
                    } ],
                    conditions: [
                    {
                        id: "always", name: "Always", dependencies: [],
                        predicate:
                        {
                            name: "always", arguments:
                            {
                            }
                        }
                    } ],
                    actions: [
                    {
                        id: "action", name: "Action", parameters:
                        {
                            result: "accepted"
                        }
                    } ],
                    rules: [
                        {
                            id: "rule-z", name: "Rule Z", event: "event", condition: "always", action: "action"
                    },
                        {
                            id: "rule-a", name: "Rule A", event: "event", condition: "always", action: "action"
                    }
                    ]
                };

                const result = evaluate
                (
                    {
                        model, event:
                        {
                            type: "event"
                        }, includeTrace: true
                    }
                );
                const permutedResult = evaluate
                (
                    {
                        model:
                        {
                            ...model, rules: [ ...model.rules ].reverse ()
                        },
                        event:
                            {
                                type: "event"
                            },
                        includeTrace: true
                    }
                );
                const untracedResult = evaluate
                (
                    {
                        model, event:
                        {
                            type: "event"
                        }, includeTrace: false
                    }
                );

                expect ( result ).toMatchObject
                (
                    {
                        outcome: "action-selected",
                        selectedAction:
                            {
                                id: "action", parameters:
                                {
                                    result: "accepted"
                                }
                            }
                    }
                );
                expect ( result.trace?.rules ).toEqual
                (
                    [
                        {
                            rule: "rule-a", condition: "always", conditionResult: true, action: "action", matched: true
                        },
                        {
                            rule: "rule-z", condition: "always", conditionResult: true, action: "action", matched: true
                        }
                    ]
                );
                expect ( permutedResult ).toEqual ( result );
                expect ( untracedResult ).toEqual
                (
                    {
                        outcome: "action-selected",
                        selectedAction: model.actions [ 0 ]
                    }
                );
            }
        );

        //-----------------------------------------------------------------------------------------------------------------
        // Test: reports deterministic ambiguity when matching rules identify distinct actions
        //
        // Description:
        //
        //   Verifies reports deterministic ambiguity when matching rules identify distinct actions and records the
        //   expected externally observable behavior for future changes.
        //
        //-----------------------------------------------------------------------------------------------------------------
        it
        (
            "reports deterministic ambiguity when matching rules identify distinct actions", () =>
            {
                const model: StatelessECAModel =
                {
                    schemaVersion: "1.0",
                    id: "ambiguous-action-selection-model",
                    name: "Ambiguous action selection model",
                    parameters: [],
                    payloads: [],
                    events: [
                    {
                        id: "signal.received", name: "Signal received"
                    } ],
                    conditions: [
                    {
                        id: "always", name: "Always", dependencies: [],
                        predicate:
                        {
                            name: "always", arguments:
                            {
                            }
                        }
                    } ],
                    actions: [
                        {
                            id: "record", name: "Record", parameters:
                        {
                        }
                    },
                        {
                            id: "alert", name: "Alert", parameters:
                        {
                        }
                    }
                    ],
                    rules: [
                        {
                            id: "record.signal", name: "Record", event: "signal.received",
                            condition: "always", action: "record"
                    },
                        {
                            id: "alert.high.2", name: "Alert two", event: "signal.received",
                            condition: "always", action: "alert"
                    },
                        {
                            id: "alert.high.1", name: "Alert one", event: "signal.received",
                            condition: "always", action: "alert"
                    }
                    ]
                };

                const result = evaluate
                (
                    {
                        model, event:
                        {
                            type: "signal.received"
                        }, includeTrace: false
                    }
                );
                const permutedResult = evaluate
                (
                    {
                        model:
                        {
                            ...model, rules: [ ...model.rules ].reverse ()
                        },
                        event:
                            {
                                type: "signal.received"
                            },
                        includeTrace: false
                    }
                );

                expect ( result ).toEqual
                (
                    {
                        outcome: "ambiguous",
                        selectedAction: null,
                        diagnostic:
                            {
                                code: "ambiguous-action-selection",
                                actionIds: [ "alert", "record" ],
                                ruleIds: [ "alert.high.1", "alert.high.2", "record.signal" ],
                                message: "Event 'signal.received' matched distinct actions [alert, record] through rules "
                                    + "[alert.high.1, alert.high.2, record.signal].",
                                pointer: "/type"
                            }
                    }
                );
                expect ( permutedResult ).toEqual ( result );
            }
        );

        //-----------------------------------------------------------------------------------------------------------------
        // Test: keeps a null selected action when a no-action result omits its trace
        //
        // Description:
        //
        //   Verifies keeps a null selected action when a no-action result omits its trace and records the expected
        //   externally observable behavior for future changes.
        //
        //-----------------------------------------------------------------------------------------------------------------
        it
        (
            "keeps a null selected action when a no-action result omits its trace", () =>
            {
                const model: StatelessECAModel =
                {
                    schemaVersion: "1.0",
                    id: "no-action-model",
                    name: "No action model",
                    parameters: [],
                    payloads: [],
                    events: [],
                    conditions: [],
                    actions: [],
                    rules: []
                };

                expect
                (
                    evaluate
                    (
                        {
                            model, event:
                            {
                                type: "unknown"
                            }, includeTrace: false
                        }
                    )
                ).toEqual
                (
                    {
                        outcome: "no-action",
                        selectedAction: null
                    }
                );
            }
        );

        //-----------------------------------------------------------------------------------------------------------------
        // Test: bounds large ambiguity messages deterministically and independently of declaration order
        //
        // Description:
        //
        //   Verifies bounds large ambiguity messages deterministically and independently of declaration order and records
        //   the expected externally observable behavior for future changes.
        //
        //-----------------------------------------------------------------------------------------------------------------
        it
        (
            "bounds large ambiguity messages deterministically and independently of declaration order", () =>
            {
                const request =
                {
                    model: createLargeAmbiguousModel ( false ),
                    event:
                    {
                        type: "signal.received"
                    },
                    includeTrace: false
                } as const;
                const originalResult = evaluate ( request );
                const repeatedResult = evaluate ( request );
                const permutedResult = evaluate
                (
                    {
                        ...request, model: createLargeAmbiguousModel ( true )
                    }
                );

                expect ( MAXIMUM_DIAGNOSTIC_MESSAGE_CODE_POINTS ).toBe ( 4096 );
                expect ( originalResult.outcome ).toBe ( "ambiguous" );

                if ( originalResult.outcome !== "ambiguous" )
                {
                    throw new Error ( "The generated evaluation must be ambiguous." );
                }

                expect ( Array.from ( originalResult.diagnostic.message ) )
                    .toHaveLength ( MAXIMUM_DIAGNOSTIC_MESSAGE_CODE_POINTS );
                expect ( originalResult.diagnostic.message.endsWith ( "\u2026" ) ).toBe ( true );
                expect ( repeatedResult ).toEqual ( originalResult );
                expect ( permutedResult ).toEqual ( originalResult );
            }
        );

        //-----------------------------------------------------------------------------------------------------------------
        // Test: truncates supplementary Unicode characters at code-point boundaries
        //
        // Description:
        //
        //   Verifies truncates supplementary Unicode characters at code-point boundaries and records the expected
        //   externally observable behavior for future changes.
        //
        //-----------------------------------------------------------------------------------------------------------------
        it
        (
            "truncates supplementary Unicode characters at code-point boundaries", () =>
            {
                const sourceMessage = "\uD83D\uDE00".repeat ( MAXIMUM_DIAGNOSTIC_MESSAGE_CODE_POINTS ) + "tail";
                const limitedMessage = limitDiagnosticMessage ( sourceMessage );

                expect ( Array.from ( limitedMessage ) ).toHaveLength ( MAXIMUM_DIAGNOSTIC_MESSAGE_CODE_POINTS );
                expect ( limitedMessage.endsWith ( "\u2026" ) ).toBe ( true );
            }
        );
    }
);
